### ISSUE TITLE
details about use your global joint position algorithm 

### ISSUE THREAD

**Issue by liwenssss** [AlextheEngineer/Ego3DHand#6]
Hi, I want to use your algorithm to get the global joint position. But I dont know how to init some parameters. Do you mind to tell me how to init the parameters including:CANVAS_DISTANCE,PXCM,INITIAL_POS,INITIAL_Y_ANGLE and INITIAL_X_ANGLE? Suppose the camera intrinsic matrix .Thank you !

**Comment by AlextheEngineer** [comment]
Hi, CANVAS_DISTANCE is less important because it's for setting the image projection behind the poses for visualization. PXCM is the ratio between pixels and cm, which should be fixed as long as the camera settings don't change and can be calculated using the 2D and 3D global coordinates provided. INITIAL_POS is the inital camera position (0,0,0) if you want the first person view. INITIAL_Y_ANGLE/INITIAL_X_ANGLE are for setting up the camera view as well. Setting the camera is for visualization only and does not affect the global joint positions. You can refer to "set_dataset" function in "pose_visualization.py" for our values for 3 different datasets. The global joint positions for Ego3DHands should already be provided. If you are trying to get the global positions from other datasets using the 2D and 3D canonical hand poses this might be applicable. Best!

**Comment by liwenssss** [comment]
Thanks for your reply!

**Comment by AlextheEngineer** [comment]
np good luck!

**Comment by liwenssss** [comment]
hello, I test and realized your canonical joint is different to my understanding, you have rotate the joints first before normalize it.I read the paper again and find your apporach is more reasonable. But I dont know how to get the rotate matrix, can you offer the code about global_to_canon?

**Comment by AlextheEngineer** [comment]
Hi here's the code I used, sorry if it's not too clean. It basically does two spherical rotation so the root joint is right in the center of the view.

[code.txt](https://github.com/AlextheEngineer/Ego3DHands/files/5899411/code.txt)

**Comment by liwenssss** [comment]
thanks again!

**Comment by liwenssss** [comment]
Hi,bother again,I test the algorithm on the [InterHand2.6M dataset](https://github.com/facebookresearch/InterHand2.6M).I first use the global camera joint get canonical joint, and I followed the step and convert the canonical joint to the global joint using 2d keypoint.But the resault is much different to the origin joint.
Is there any specific request to use the algorithm?

**Comment by AlextheEngineer** [comment]
Hi, can you tell me more about how it seems much different? Have you added this dataset's info in the "set_dataset" function of "pose_visualization.py"? I would also make sure the data input format is exactly the same as ours. For example, our 2D joint annotations are in form of percentage so it's adapts to different image sizes. Also, both the 2D and 3D joint annotations of our data follow the order of row (vertical), column (horizontal) and then depth (many datasets use (x,y,z) where x and y might be referring to the column (horizontal) and row (vertical) instead.)

**Comment by liwenssss** [comment]
> Hi, can you tell me more about how it seems much different? Have you added this dataset's info in the "set_dataset" function of "pose_visualization.py"? I would also make sure the data input format is exactly the same as ours. For example, our 2D joint annotations are in form of percentage so it's adapts to different image sizes. Also, both the 2D and 3D joint annotations of our data follow the order of row (vertical), column (horizontal) and then depth (many datasets use (x,y,z) where x and y might be referring to the column (horizontal) and row (vertical) instead.)

I use the code from [here](https://github.com/facebookresearch/InterHand2.6M/blob/75a2d29421ad9dd8607e7948dcfbccf706b1eef0/data/InterHand2.6M/dataset.py#L146) to get data. I use the global 3d joint gt_joint_coord  and 2.5d joint pred_joint_coord_img , in my test case pred_joint_coord_img  is the gt 2.5d joint.
First I follow the step get the 2d keypoint in the original image space by inverse affine transformation:
```
pred_joint_coord_img[:, 0] = pred_joint_coord_img[:, 0] / cfg.output_hm_shape[2] * cfg.input_img_shape[0]
pred_joint_coord_img[:, 1] = pred_joint_coord_img[:, 1] / cfg.output_hm_shape[1] * cfg.input_img_shape[1]
for j in range(42):
    pred_joint_coord_img[j, :2] = trans_point2d(pred_joint_coord_img[j, :2], inv_trans[n])
```
then I make it in the form of percentage:
```
pred_joint_coord_img[:, 0] = pred_joint_coord_img[:, 0] / 334
pred_joint_coord_img[:, 1] = pred_joint_coord_img[:, 1] / 512
```
I choose the right hand to test and change the coordinates order:
```
def uv2vu(uv):
    res = np.concatenate((uv[:,1,None],uv[:,0,None]),1)
    return res

kp_right = pred_joint_coord_img[:21,:2]
kp_right = uv2vu(kp_right)
```
I use the 3d global joint to get the canonical joint:
```
def xyz2yxz(xyz):
    res = np.concatenate((xyz[:,1,None],xyz[:,0,None],xyz[:,2,None]),1)
    return res

gt_cam_right = gt_joint_coord[:21]
gt_cam_left = gt_joint_coord[21:]
pred_cam_right = xyz2yxz(gt_cam_right)
can_right, bone_l = global_to_canon(pred_cam_right)
```
and I set the parameters:
```
gc.CAMERA_FOCAL_LENGTH = 0.999 # use a random focal length
pxcm = gc.calculate_pxcm(pos_2d=kp_right, glob=pred_cam_right)
gc.PXCM = pxcm
gc.GLOBAL_HAND_LENGTH = bone_l
pred_right = gc.calculate_global_positions(hand_2d=kp_right, hand_3d=can_right, hand_side=1)
pred_right = xyz2yxz(pred_right) # back to the origin order
```
finally , I print the pred_right and gt_cam_right , they are much more different.Can you give any advice?Thanks!

**Comment by AlextheEngineer** [comment]
I would investigate closer by looking at the actual values and shapes of the 2D joint locations, 3D canonical joint locations, 3D global joint locations, and see if they roughly match the image of the hand pose. If those all match, then the issue is probably in the dataset-specific values. If you can't find obvious error can you provide a zip file with the information I mentioned?

**Comment by liwenssss** [comment]
Can I email you an example data for you so that you can test it for further investigation?

**Comment by AlextheEngineer** [comment]
Can you send me the data I mentioned? I'm sure if the data and the parameters are properly set it can reconstruct the global 3D joint locations. By just looking at the values and the image I might be able to find something wrong.

**Comment by liwenssss** [comment]
[test.zip](https://github.com/AlextheEngineer/Ego3DHands/files/5901520/test.zip)
here is 4 sample data and a script to test and show, thanks for your help!

**Comment by AlextheEngineer** [comment]
I see the global 3d joint locations is "joint_cam.npy", however I can't find the 2d joint locations in percentage form and the canonical 3d joint locations. Also, are the joint ordering the same as ours? 0-wrist, 1-4 thumb, 5-8 index, 9-12 middle, 13-16 ring, 17-20 pinky?

**Comment by liwenssss** [comment]
yes, you can run the test.py and see it .I use your global_to_canon to get canonical  3d joint locations, and I get 2d joint locations in percentage form in in test.py.

**Comment by AlextheEngineer** [comment]
I took a quick look and for image 0, the 2D joint locations don't seem to match the image. For example, the first joint (wrist) is at the bottom right of the image, but its value is [0.51807934  0.75469774], which indicates a row value approximately in the middle and a column value 3/4 from the left. There might be other errors but I recommend running the sample code and try to match your data to exactly the same format in our example and verify the input data is correct (shapes, values, joint/xyz ordering)

**Comment by liwenssss** [comment]
The image I upload the the cropped one.The original image is here:
![0](https://user-images.githubusercontent.com/11902259/106413453-0f984f00-6485-11eb-88cd-e2d348c74bfb.png)
it seems the root joint is at the middle row and nearly at the most right of the image.
I debug the code but I cant find where the error is. If I cant solve it,maybe I will try another method, Thanks for your answer.

**Comment by AlextheEngineer** [comment]
I've investigated and found that the calculated PXCM varies significantly when calculating using different joints. For example, 
when using joint 7
PXCM = 6736.86102587 pixels per cm
when using joint 8
PXCM = 222.315107031 pixels per cm
This is not supposed to happen as PXCM should be the same for the same camera setting. As it is calculated using the 3d global joint locations and the 2d joint locations (focal length is irrelevant because it cancels in the overall calculation.) I've looked into why this happens and this is what I found for joint 7 and 8:
![106413453-0f984f00-6485-11eb-88cd-e2d348c74bfb](https://user-images.githubusercontent.com/11249868/106422896-a250d380-641c-11eb-8d99-5c730b85959a.png)
You can see that joint 7 and 8 has the following 2D annotations which I've plotted on the graph (row first, column second).
[ 136.87321472  181.94732785]
[ 120.77779388  163.6286183 ]
On the other hand, joint 7 and 8 has the following 3D global annotations:
[ -10.16983128   -0.26921931  109.20505524]
[ -11.45845222   -1.82547915  108.33790588]
The problem is that you can see from camera view, joint 7 and 8 should have positive and negative value for its 3D horizontal world coordinate respectively. However, both 7 and 8 have negative 3D horizontal world coordinate (-0.26921931 and -1.82547915). This results in a huge difference in the calculated PXCM (6736.86102587 and 222.315107031.) Also note that joint 7 should be horizontally further from the green line but joint 7 has a smaller 3D horizontal value.
In summary, although small, the error in their 3D global ground truth annotation results in error when trying to perfectly reconstruct through projection. However I find this odd because the 2D joint locations should be obtained using the camera intrisics matrix and the 3D global joint locations. Is that how you obtained the 2D joint locations?

**Comment by liwenssss** [comment]
This is funny.I get the 2d keypoint by intrinsic matrix(you can see it in  focal.npy and princpt.npy I upload).I guess maybe it is because the provided fx and fy are different.Maybe these two parameters make the algorithm get the wrong answer.

**Comment by liwenssss** [comment]
> This is funny.I get the 2d keypoint by intrinsic matrix(you can see it in focal.npy and princpt.npy I upload).I guess maybe it is because the provided fx and fy are different.Maybe these two parameters make the algorithm get the wrong answer.

I totally follow the step provided by the dataset [InterHand2.6M](https://github.com/facebookresearch/InterHand2.6M/blob/master/data/InterHand2.6M/dataset.py#L146).However, in this dataset,some joints are not valid. I choose the data which has all the joints valid to test. I guess  maybe I cant use your algorithm test on this dataset.

**Comment by AlextheEngineer** [comment]
This is one issue, the other one is that their 3D ground truth world coordinate has 2 negative values for 2 joints that are on 2 different sides of the center line. This is definitely annotation error (although it might be too small and thus trivial). However, the inconsistency between the 2D and 3D is the key.

**Comment by AlextheEngineer** [comment]
I think the fact that you have joint 7 with 3d global coordinate of [ -10.16983128   -0.26921931  109.20505524] and 2d image coordinate of [ 136.87321472  181.94732785] is very odd. The 3D location indicates the joint should be on left half of the view (-0.269 < 0) but its 2D says it's on the right half (182 > 167). From the image, we can clearly see that joint 7 is on the right half of the image, which means the 3D world coordinate is inaccurate. But how did you compute an accurate 2D using an inaccurate 3D and the camera intrinsics matrix? If you can make them consistent and solve the problem of very inconsistent PXCM (they should be approximately the same for all joints), my algorithm should work for you. Best!

**Comment by liwenssss** [comment]
Thanks for your answer~

**Comment by AlextheEngineer** [comment]
np, I'm very curious how this can happen. The only thing I can think of is maybe somehow you are using the right 2D coordinates but messed up on pairing its global 3d coordinates ground truth (so the 3D global joint coordinates you provided is not the one for the 2D coordinates). I don't see how using camera intrinsics you can get a joint that was originally on the left side in 3D to end up on the right side in 2D.

**Comment by liwenssss** [comment]
Hi, I get a nearly right resault by the following way:
I first  get the canonical pose.
```
gt_cam_right = gt_cam[:,:21]
# gt_yxz = xyz2yxz(gt_cam_right)
gt_can, bone_l = global_to_canon(gt_cam_right)
```
Than I get the percentage value of the keypoint with the princpt value instead of the image size:
```
kp_right[:, 0] = kp_right[:, 0] / (pp[0] * 2)
kp_right[:, 1] = kp_right[:, 1] / (pp[1] * 2)
```
There are two difference:
- I use the princpt value to get the  percentage value of the keypoint instead of the image size
- I set X=0,Y=1 without change the rotation matrices. I have not find where to change now.
I use the following code to test:
```
mpjpe = []
for j in range(21):
    mpjpe.append(np.sqrt(np.sum((gt_cam_right[j] - pred_right[j]) ** 2)))
print(np.mean(mpjpe))
```
the final error is 0.02mm(the joint coord I input is in mm).

**Comment by AlextheEngineer** [comment]
Good to know it works! Changing X and Y saves you the effort to swap the x, y on the coordinates. However I'm unsure why obtaining the 2D percentage value using the princpt fixes the problem since I checked the 2D percentage value obtained using the image size and they seem correct.

**Comment by liwenssss** [comment]
Hi, In the paper you said "However, for the task of 3D global hand pose estimation in the next stage, we replace p from P oseNet2D with the (x, y) component of wcan projected back into the pixel space since the consistency between 2D and 3D poses is important for our projection algorithm". How do you do the project step ? I mean how do you get the camera matrix between the canonical joint position and the predicted 2d keypoint?

**Comment by AlextheEngineer** [comment]
I basically compute the 2D pixel distance between wrist and mMCP and use this distance to scale the 3D canonical pose's (x, y) component. Then I shift the scaled (x, y) component to the same 2D location of the mMCP (since mMCP is the root joint, the 3D canonical pose is zero-centered on it). After that the (x, y) component of the scaled and shifted 3D canonical pose is used as the 2D pose instead of the originally estimated 2D joint locations.
[code.txt](https://github.com/AlextheEngineer/Ego3DHands/files/5922320/code.txt)

**Comment by liwenssss** [comment]
“Lproj indicates the MSE between the (x, y) component of wcan and p projected into the same canonical frame”， does it mean that this step use 2d keypoint p to get the 3d canonical joint location is a inverse step of that one?

**Comment by AlextheEngineer** [comment]
Yes pretty much, except I don't use only the key bone for the scaling, but the average distance from all joints to the wrist. Here's the code snippet. You can ignore the absent hand part, absent hand can be checked after the detection network and therefore should be able to skip the 2D/3D estimation parts.
[code.txt](https://github.com/AlextheEngineer/Ego3DHands/files/5922546/code.txt)

**Comment by liwenssss** [comment]
> I basically compute the 2D pixel distance between wrist and mMCP and use this distance to scale the 3D canonical pose's (x, y) component. Then I shift the scaled (x, y) component to the same 2D location of the mMCP (since mMCP is the root joint, the 3D canonical pose is zero-centered on it). After that the (x, y) component of the scaled and shifted 3D canonical pose is used as the 2D pose instead of the originally estimated 2D joint locations.
> [code.txt](https://github.com/AlextheEngineer/Ego3DHands/files/5922320/code.txt)

I test the code as follow wich kp_right is the percentage of right hand keypoint:
```
 kpts_3d_0 = project_2d_from_3d(kp_right, gt_can_right, MIDDLE_FINGER_1)
```
and I print kpts_3d_0 and kp_right:
```
kpts_3d_0:
[[0.67926214 0.24429576]
 [0.68799089 0.31279283]
 [0.69980048 0.37405001]
 [0.6876113  0.46666673]
 [0.46399403 0.21486738]
 [0.50423407 0.24951249]
 [0.54393679 0.2883378 ]
 [0.60275265 0.35681811]
 [0.40874777 0.22660836]
 [0.45465793 0.26886002]
 [0.50146762 0.31670364]
 [0.56617063 0.39485279]
 [0.4055495  0.2884332 ]
 [0.44607929 0.33102961]
 [0.48839191 0.37482349]
 [0.55430725 0.44026527]
 [0.42955312 0.37750213]
 [0.46575065 0.40964061]
 [0.49800039 0.43907857]
 [0.55157452 0.48810001]
 [0.66564099 0.54355004]]
kp_right:
[[0.69816405 0.26738164]
 [0.7082078  0.32540467]
 [0.72106415 0.37728038]
 [0.7071386  0.45520505]
 [0.44203204 0.23639382]
 [0.49151886 0.26789677]
 [0.53960323 0.30259675]
 [0.6092299  0.36237085]
 [0.37367684 0.24581918]
 [0.43078712 0.28403753]
 [0.4882687  0.32668483]
 [0.56617063 0.3948528 ]
 [0.36943766 0.30047092]
 [0.4196502  0.33848566]
 [0.47179663 0.37728924]
 [0.55194235 0.434347  ]
 [0.39786986 0.3794557 ]
 [0.44310188 0.40806392]
 [0.48308334 0.43398026]
 [0.54854476 0.4764972 ]
 [0.68094814 0.5191768 ]]
```
and I calculate the error:
```
pred_right = gc.calculate_global_positions(hand_2d=kpts_3d_0, hand_3d=gt_can_right, hand_side=1)

mpjpe = []
for j in range(21):
    mpjpe.append(np.sqrt(np.sum((gt_cam_right[j] - pred_right[j]) ** 2)))
print(np.mean(mpjpe))
```
the error is over 100. am I do sth wrong?

**Comment by AlextheEngineer** [comment]
I'm not exactly sure, but I used wrist index instead of mMCP index, the ratio might have larger fluctuation when you use mMCP because many finger joints are very close to it. Also if the error is in mm, 100mm is 10cm, which isn't too bad since we are trying to get global positions using only 2D data (move your hand back and forth 10cm, there can be very little visual differences.)

**Comment by liwenssss** [comment]
I tried , the error is still large

**Comment by AlextheEngineer** [comment]
but using "kp_right" instead of "kpts_3d_0" gets small error?

**Comment by liwenssss** [comment]
Yes, as the previous discussion, I got a naerly right result.
When I run the code:
```
kpts_3d_r = project_2d_from_3d(kp_right, gt_can_right, 20)
pred_right = gc.calculate_global_positions(hand_2d=kpts_3d_r, hand_3d=gt_can_right, hand_side=1)
mpjpe = []
for j in range(21):
    mpjpe.append(np.sqrt(np.sum((gt_cam_right[j] - pred_right[j]) ** 2)))
print(np.mean(mpjpe))
```
I got error:30.96545627926693. However,when I run the code:
```
pred_right = gc.calculate_global_positions(hand_2d=kp_right, hand_3d=gt_can_right, hand_side=1)
mpjpe = []
for j in range(21):
    mpjpe.append(np.sqrt(np.sum((gt_cam_right[j] - pred_right[j]) ** 2)))
print(np.mean(mpjpe))
```
I got the error:0.02876510094541182.
In my case, the data as follows:
```
kp_right:
[[0.69816405 0.26738164]
 [0.7082078  0.32540467]
 [0.72106415 0.37728038]
 [0.7071386  0.45520505]
 [0.44203204 0.23639382]
 [0.49151886 0.26789677]
 [0.53960323 0.30259675]
 [0.6092299  0.36237085]
 [0.37367684 0.24581918]
 [0.43078712 0.28403753]
 [0.4882687  0.32668483]
 [0.56617063 0.3948528 ]
 [0.36943766 0.30047092]
 [0.4196502  0.33848566]
 [0.47179663 0.37728924]
 [0.55194235 0.434347  ]
 [0.39786986 0.3794557 ]
 [0.44310188 0.40806392]
 [0.48308334 0.43398026]
 [0.54854476 0.4764972 ]
 [0.68094814 0.5191768 ]]
```
```
gt_can_right:
[[ 0.57605197 -0.76688936  0.30482442]
 [ 0.62051343 -0.41798718  0.30619272]
 [ 0.68066769 -0.10596268  0.38289866]
 [ 0.6185799   0.36579724  0.33998686]
 [-0.52045491 -0.91678809 -0.30145039]
 [-0.31548504 -0.74031697 -0.18935575]
 [-0.11325208 -0.54255329 -0.06724125]
 [ 0.18633712 -0.19373645  0.14639621]
 [-0.80186173 -0.8569833  -0.40198854]
 [-0.56801001 -0.64176687 -0.30352611]
 [-0.32957647 -0.39806678 -0.18831175]
 [ 0.          0.          0.        ]
 [-0.81815267 -0.54206733 -0.43953628]
 [-0.61170689 -0.32509486 -0.38892813]
 [-0.39617991 -0.10202282 -0.31465978]
 [-0.06042826  0.23131667 -0.1531614 ]
 [-0.69588589 -0.08837871 -0.53871951]
 [-0.51150729  0.07532446 -0.46802219]
 [-0.3472374   0.22527201 -0.39224852]
 [-0.07434795  0.47497153 -0.26337839]
 [ 0.50667017  0.75741622  0.4118325 ]]
```
```
gt_cam_right:
[[  64.9601    -105.256096  1127.7513   ]
 [  68.321686   -79.08162   1128.8938   ]
 [  72.968025   -55.91258   1135.5486   ]
 [  68.27297    -20.379707  1133.9119   ]
 [ -18.254791  -114.58452   1083.379    ]
 [  -2.692193  -101.69831   1092.0505   ]
 [  12.680658   -87.24619   1101.545    ]
 [  35.504734   -61.74992   1118.2428   ]
 [ -39.525585  -109.77906   1076.4325   ]
 [ -21.811712   -93.94257   1084.1622   ]
 [  -3.7279084  -76.02192   1093.2322   ]
 [  21.314209   -46.750175  1108.1426   ]
 [ -40.7838     -86.03147   1074.6357   ]
 [ -25.196636   -69.91177   1078.8207   ]
 [  -8.893104   -53.409294  1084.7867   ]
 [  16.570047   -28.909304  1097.4719   ]
 [ -31.718925   -51.676434  1068.4542   ]
 [ -17.762928   -39.617214  1074.0095   ]
 [  -5.3103     -28.60617   1079.9312   ]
 [  15.380687   -10.278586  1089.9948   ]
 [  59.997757     8.776844  1140.7024   ]]
```
in my case,WRIST = 20,MIDDLE_FINGER_1 = 11,PINKY_1 = 19

**Comment by AlextheEngineer** [comment]
So "project_2d_from_3d" probably doesn't reconstruct the 2D joint locations perfectly (if it's not the same as the ground truth 2D joint locations). 30mm is actually not bad for global estimation because estimating the absolute 3D distance using 2D data is very challenging.
